### PR TITLE
Make the glue compute short circuit only if possible

### PR DESCRIPTION
**Pull request by ezyang**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #94512
* #94473
* __->__ #94437
* #94431

If the inputs are unhinted, they will use the branchless implementation.

Signed-off-by: Edward Z. Yang <ezyang@meta.com>

### CODE REVIEW

File: c10/core/TensorImpl.cpp [hunk 1]
```diff
@@ -560,56 +560,82 @@ SymBool TensorImpl::compute_non_overlapping_and_dense(identity<SymBool>) const {
...
-          __FILE__, __LINE__) ||
-      compute_non_overlapping_and_dense(type_id).guard_bool(__FILE__, __LINE__);
+  if (definitely_true(extra_meta_->is_contiguous_)) {
+    return true;
+  }
+  if (definitely_true(extra_meta_->is_channels_last_contiguous_)) {
+    return true;
+  }
+  return extra_meta_->is_contiguous_ |
+      extra_meta_->is_channels_last_contiguous_ |
```
Comment (bdhirsh) [inline c10/core/TensorImpl.cpp:586]: hmm is there a reason for the change from logical or to bitwise or? (Do we not want to use || on SymBools?)

Comment (bdhirsh) [reply c10/core/TensorImpl.cpp:586]: Oh, from looking through the next PR, it looks like our python sympy impl eventually calls `eval_is_non_overlapping_and_dense()`, which treats the output as an int.

Comment (ezyang) [reply c10/core/TensorImpl.cpp:586]: Logical or is not overrideable, bitwise or goes into `sym_or` which is branchless.

**Review by voznesenskym — APPROVED** [review]

**Pull request closed without merging**